Rpackage/R/fitAndScore.R

```diff
@@ -0,0 +1,191 @@
+                                       maxdepth = as.numeric(parameters$maxdepth), margin = as.numeric(parameters$margin), 
+                                       loss = loss, min_sample = as.numeric(parameters$min_sample))
+  }
+  
+  ### master tree
+  master_tree <- mmit(target.mat, feature.mat,  
+                      maxdepth = as.numeric(parameters$maxdepth), margin = as.numeric(parameters$margin), 
+                      loss = loss, min_sample = as.numeric(parameters$min_sample))
+  
+  
+  ### if pruning
+  if(pruning){
+    # Get the pruned master and cross-validation trees
+    master_data <- pruning(master_tree)
+    master_alphas <- master_data[, 1]
+    master_pruned_trees <- master_data[, -1]
+    
+    fold_alphas <- NULL
+    fold_prune_trees <- NULL
+    for(i in 1 : n_folds){
+      fold_data <- pruning(fold_tree[[i]])
+      fold_alphas[[i]] <- fold_data[, 1] 
+      fold_prune_trees[[i]] <- fold_data[, -1]
+    }
+    
+    ### alphas list should not contain repeating alpha
+    master_alphas <- master_alphas[!duplicated(master_alphas)]
+    fold_alphas <- fold_alphas[!duplicated(fold_alphas)]
+    
+    
+    # Compute the test risk for all pruned trees of each fold
+    alpha_path_score <- NULL
+    for(i in 1 : n_folds){
+      for(j in 1 : length(fold_prune_trees[[i]][,1])){
+        ### convert pruned tree list to partynode
+        node <- fold_prune_trees[[i]][[j]]
+        
+        fit <- fitted_node(node, feature.mat[fold_split_idx$test[i,],])
+        n <- nodeapply(node, ids = fit, info_node)
+        ###prediction of test data
+        prediction <- matrix(unlist(n), nrow = length(n), byrow = T)[,1]
+        ###error calc
+        fold_test_scores <- scorer(target.mat[fold_split_idx$test[i,],], prediction)
+        
+        ### creating a dataframe with init alpha, final alpha, score value
+        if(j < (length(fold_alphas[[i]]) - 1)){
+          alpha_path_score <- rbind(alpha_path_score, c(fold_alphas[[i]][j], fold_alphas[[i]][j+1], fold_test_scores))
+        }
+        else{
+          alpha_path_score <- rbind(alpha_path_score, c(fold_alphas[[i]][j], Inf, fold_test_scores))
+        }
+        colnames(alpha_path_score) <- c("init alpha", " final alpha", "score")
+        alpha_path_score <- as.data.frame(alpha_path_score)
+      }
+      
+    }
+    
+    # Prune the master tree based on the CV estimates
+    alphas <- NULL
+    alpha_cv_scores <- NULL
+    alpha_train_scores <- NULL
+    alpha_train_objective_values <- NULL
+    best_alpha <- -1
+    best_score <- Inf
+    best_tree <- NULL
+    
+    for(i in 1:length(master_pruned_trees[,1])){
+      if(i < (length(master_alphas) - 1)){
+        geo_mean_alpha_k <- sqrt(master_alphas[[i]]* master_alphas[[i + 1]])
+      }
+      else{
+        geo_mean_alpha_k <- Inf
+      }
+      
+      ### compute cv_score as mean of each fold scores
+      cv_score <- 0
+      for(i in 1:n_folds){
```

Comment:
`i` does not even appear in this list. I really don't understand what's going on here.

Reply:
alpha_path_score should be indexed with the fold